src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -0,0 +1,945 @@
+          assert(nnz_each_row_tmp[i] == kRowPtr_[i+1]-kRowPtr_[i]-1);
+          nnz_tmp = nnz_each_row_tmp[i] + kRowPtr_[i];
+          jCol_[nnz_tmp] = i;
+          kVal_[nnz_tmp] = M_->M()[M_->numberOfNonzeros()-n_+i];
+          index_covert_CSR2Triplet_[nnz_tmp] = M_->numberOfNonzeros()-n_+i;
+          total_nnz_tmp += 1;
+
+          std::vector<int> ind_temp(kRowPtr_[i+1]-kRowPtr_[i]);
+          std::iota(ind_temp.begin(), ind_temp.end(), 0);
+          std::sort(ind_temp.begin(), ind_temp.end(),[&](int a, int b){ return jCol_[a+kRowPtr_[i]]<jCol_[b+kRowPtr_[i]]; });
+
+          reorder(kVal_+kRowPtr_[i],ind_temp,kRowPtr_[i+1]-kRowPtr_[i]);
+          reorder(index_covert_CSR2Triplet_+kRowPtr_[i],ind_temp,kRowPtr_[i+1]-kRowPtr_[i]);
+          std::sort(jCol_+kRowPtr_[i],jCol_+kRowPtr_[i+1]);
+        }
+      }
+      delete [] nnz_each_row_tmp;
+    }
+
+
+    /*
+     * initialize KLU and cuSolver parameters
+     */
+
+    //handles
+    cusparseCreate(&handle_); 
+    cusolverSpCreate(&handle_cusolver_);
+    cublasCreate(&handle_cublas_);
+
+    //descriptors
+    cusparseCreateMatDescr(&descr_A_);
+    cusparseSetMatType(descr_A_, CUSPARSE_MATRIX_TYPE_GENERAL);
+    cusparseSetMatIndexBase(descr_A_, CUSPARSE_INDEX_BASE_ZERO);
+
+    cusparseCreateMatDescr(&descr_M_);
+    cusparseSetMatType(descr_M_, CUSPARSE_MATRIX_TYPE_GENERAL);
+    cusparseSetMatIndexBase(descr_M_, CUSPARSE_INDEX_BASE_ZERO);
+
+    //info (data structure where factorization is stored)
+    cusolverSpDestroyCsrluInfoHost(info_lu_);
+    cusolverSpDestroyGluInfo(info_M_);
+    cusolverSpCreateCsrluInfoHost(&info_lu_);
+    cusolverSpCreateGluInfo(&info_M_);
+
+    //KLU 
+    klu_defaults(&Common_) ;
+
+    //modify as you please
+    //TODO: consider making a part of setup options that can be called from a user side
+    Common_.btf = 0;
+    Common_.ordering = 1; //COLAMD; use 0 for AMD
+    Common_.tol = 0.1;
+    Common_.scale = -1;
+    Common_.halt_if_singular=0;
+    //allocate gpu data
+    free(mia_); 
+    free(mja_);
+    mia_ = NULL; 
+    mja_ = NULL;
+    cudaFree(devx_);
+    cudaFree(devr_);
+    devx_ = NULL;
+    devr_ = NULL;
+    checkCudaErrors(cudaMalloc(&devx_, n_ * sizeof(double)));
+    checkCudaErrors(cudaMalloc(&devr_, n_ * sizeof(double)));
+    this->newKLUfactorization();
+  }
+
+  //helper private function needed for format conversion
+
+  int hiopLinSolverIndefSparseCUSOLVER::createM(const int n, 
+                                                const int nnzL, 
+                                                const int* Lp, 
+                                                const int* Li,
+                                                const int nnzU, 
+                                                const int* Up, 
+                                                const int* Ui){
+
+    int row;
+    for(int i = 0;i<n;++i) {
+      //go through EACH COLUMN OF L first
+      for(int j = Lp[i]; j<Lp[i+1]; ++j) {
```

Comment:
I see this exact code used four times in this PR - can we pull it out into a function?